src/eos-updater-branch-file.c

```diff
@@ -0,0 +1,309 @@
+                            error);
+  if (info == NULL)
+    return NULL;
+
+  mod_time = g_file_info_get_attribute_uint64 (info,
+                                               G_FILE_ATTRIBUTE_TIME_MODIFIED);
+  return g_date_time_new_from_unix_utc (mod_time);
+}
+
+static gboolean
+read_file_to_bytes (GFile *file,
+                    GCancellable *cancellable,
+                    GBytes **out_bytes,
+                    GError **error)
+{
+  g_autofree gchar *contents = NULL;
+  gsize len = 0;
+
+  if (!g_file_query_exists (file, cancellable))
+    return TRUE;
+
+  if (!g_file_load_contents (file,
+                             cancellable,
+                             &contents,
+                             &len,
+                             NULL,
+                             error))
+    return FALSE;
+
+  *out_bytes = g_bytes_new_take (g_steal_pointer (&contents), len);
+  return TRUE;
+}
+
+EosBranchFile *
+eos_branch_file_new_empty (void)
+{
+  return g_object_new (EOS_TYPE_BRANCH_FILE, NULL);
+};
+
+// TODO: put it in util header
+static GFile *
+get_eos_extensions_dir (OstreeRepo *repo)
+{
+  g_autofree gchar *rel_path = g_build_filename ("extensions", "eos", NULL);
+
+  return g_file_get_child (ostree_repo_get_path (repo), rel_path);
+}
+
+static void
+get_branch_file_paths (GFile *ext_path,
+                       GFile **branch_file,
+                       GFile **branch_file_sig)
+{
+  *branch_file = g_file_get_child (ext_path, "branch_file");
+  *branch_file_sig = g_file_get_child (ext_path, "branch_file.sig");
+}
+
+EosBranchFile *
+eos_branch_file_new_from_repo (OstreeRepo *repo,
+                               GCancellable *cancellable,
+                               GError **error)
+{
+  g_autoptr(GFile) ext_dir = get_eos_extensions_dir (repo);
+  g_autoptr(GFile) branch_file = NULL;
+  g_autoptr(GFile) signature = NULL;
+
+  get_branch_file_paths (ext_dir, &branch_file, &signature);
+  return eos_branch_file_new_from_files (branch_file, signature, cancellable, error);
+}
+
+EosBranchFile *
+eos_branch_file_new_from_files (GFile *branch_file,
+                                GFile *signature,
+                                GCancellable *cancellable,
+                                GError **error)
+{
+  g_autoptr(GBytes) branch_file_bytes = NULL;
+  g_autoptr(GBytes) signature_bytes = NULL;
+  g_autoptr(GDateTime) download_time = NULL;
+
+  g_return_val_if_fail (G_IS_FILE (branch_file), NULL);
+  g_return_val_if_fail (G_IS_FILE (signature), NULL);
+
+  if (!read_file_to_bytes (branch_file,
+                           cancellable,
+                           &branch_file_bytes,
+                           error))
+    return NULL;
+  if (branch_file_bytes == NULL)
+    {
+      g_autofree gchar *path = g_file_get_path (branch_file);
+
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                   "cached branch file in %s not found", path);
+      return NULL;
+    }
```

Comment:
See above - maybe it would be better to just propagate the error that GIO already sets.